includes/utils/class-amp-image-dimension-extractor.php

```diff
@@ -150,6 +151,125 @@ private static function register_callbacks() {
 		do_action( 'amp_extract_image_dimensions_batch_callbacks_registered' );
 	}
 
+	/**
+	 * Extract dimensions from filename if dimension exists or from file system.
+	 *
+	 * @param array $dimensions Image urls mapped to dimensions.
+	 * @return array Dimensions mapped to image urls, or false if they could not be retrieved
+	 */
+	public static function extract_by_filename_or_filesystem( $dimensions ) {
+
+		if ( empty( $dimensions ) || ! is_array( $dimensions ) ) {
+			return [];
+		}
+
+		$using_ext_object_cache = wp_using_ext_object_cache();
+		$ext_types              = wp_get_ext_types();
+		if ( empty( $ext_types['image'] ) ) {
+			return $dimensions;
+		}
+		$image_ext_types = $ext_types['image'];
+		unset( $ext_types );
+
+		$upload_dir      = wp_get_upload_dir();
+		$base_upload_uri = trailingslashit( $upload_dir['baseurl'] );
+		$base_upload_dir = trailingslashit( $upload_dir['basedir'] );
+
+		foreach ( $dimensions as $url => $value ) {
+
+			// Check whether some other callback attached to the filter already provided dimensions for this image.
+			if ( ! empty( $value ) && is_array( $value ) ) {
+				continue;
+			}
+
+			$url_without_query_fragment = strtok( $url, '?#' );
+
+			// Parse info out of the URL, including the file extension and (optionally) the dimensions.
+			if ( ! preg_match( '/(?:-(?<width>[1-9][0-9]*)x(?<height>[1-9][0-9]*))?\.(?<ext>\w+)$/', $url_without_query_fragment, $matches ) ) {
+				continue;
+			}
+
+			// Skip images don't have recognized extensions.
+			if ( ! in_array( $matches['ext'], $image_ext_types ) ) {
+				continue;
+			}
+
+			// Use image dimension from the file name.
+			if ( ! empty( $matches['width'] ) && ! empty( $matches['height'] ) ) {
+				$dimensions[ $url ] = [
+					'width'  => (int) $matches['width'],
+					'height' => (int) $matches['height'],
+				];
+				continue;
+			}
+
+			// Verify that the URL is for an uploaded file.
+			if ( 0 !== strpos( $url_without_query_fragment, $base_upload_uri ) ) {
+				continue;
+			}
+			$upload_relative_path = substr( $url_without_query_fragment, strlen( $base_upload_uri ) );
+
+			// Bail if the URL contains relative paths.
+			if ( 0 !== validate_file( $upload_relative_path ) ) {
+				continue;
+			}
```

Comment:
I think these should be the first set of checks before parsing the URL for dimensions.

Reply:
I meant moving both checks. So it would look like:

```diff
diff --git a/includes/utils/class-amp-image-dimension-extractor.php b/includes/utils/class-amp-image-dimension-extractor.php
--- a/includes/utils/class-amp-image-dimension-extractor.php	(revision 35310f1ff829be5a145d14e72af27a28b2cfc0c6)
+++ b/includes/utils/class-amp-image-dimension-extractor.php	(date 1628092492414)
@@ -184,6 +184,17 @@
 
 			$url_without_query_fragment = strtok( $url, '?#' );
 
+			// Verify that the URL is for an uploaded file.
+			if ( 0 !== strpos( $url_without_query_fragment, $base_upload_uri ) ) {
+				continue;
+			}
+			$upload_relative_path = substr( $url_without_query_fragment, strlen( $base_upload_uri ) );
+
+			// Bail if the URL contains relative paths.
+			if ( 0 !== validate_file( $upload_relative_path ) ) {
+				continue;
+			}
+
 			// Parse info out of the URL, including the file extension and (optionally) the dimensions.
 			if ( ! preg_match( '/(?:-(?<width>[1-9][0-9]*)x(?<height>[1-9][0-9]*))?\.(?<ext>\w+)$/', $url_without_query_fragment, $matches ) ) {
 				continue;
@@ -202,17 +213,6 @@
 				];
 				continue;
 			}
-
-			// Verify that the URL is for an uploaded file.
-			if ( 0 !== strpos( $url_without_query_fragment, $base_upload_uri ) ) {
-				continue;
-			}
-			$upload_relative_path = substr( $url_without_query_fragment, strlen( $base_upload_uri ) );
-
-			// Bail if the URL contains relative paths.
-			if ( 0 !== validate_file( $upload_relative_path ) ) {
-				continue;
-			}
 
 			// Get image dimension from file system.
 			$image_file = $base_upload_dir . $upload_relative_path;
```

Reply:
Ah, I just realized that this needs to be ordered in this way. The reason is that we will use the dimensions in the filename for remote files which have URLs that don't start with `$base_upload_uri`. 

So the order is correct as it stands.

Reply:
Ah I see, makes sense 👍